### PR TITLE
Simplify client/OOP asset sync more.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/59218

Followup to https://github.com/dotnet/roslyn/pull/60761.

### CODE REVIEW

File: src/Workspaces/Remote/Core/SolutionAssetStorage.cs [hunk 1]
```diff
@@ -35,15 +33,18 @@ internal partial class SolutionAssetStorage
         /// the same storage here so that all OOP calls can safely call back into us and get the assets they need, even
         /// if individual calls get canceled.
         /// </summary>
-        private readonly Dictionary<Checksum, (int refCount, Scope scope)> _checksumToRefCountedScope = new();
+        private readonly Dictionary<Checksum, Scope> _checksumToScope = new();
 
-        /// <summary>
-        /// Map from solution checksum scope id to its associated <see cref="SolutionState"/>.
-        /// </summary>
-        private readonly ConcurrentDictionary<int, (SolutionState Solution, SolutionReplicationContext ReplicationContext)> _solutionStates = new(concurrencyLevel: 2, capacity: 10);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/SolutionAssetStorage.cs:43]: merged these dictionaries into one dictionary.  Merged RefCount into Scope itself.

File: src/Workspaces/Remote/Core/SolutionAssetStorage.cs [hunk 2]
```diff
@@ -148,9 +135,11 @@ private void DisposeScope(Scope scope)
         }
 
         /// <summary>
-        /// Retrieve assets of specified <paramref name="checksums"/> available within <paramref name="scopeId"/> scope from the storage.
+        /// Retrieve assets of specified <paramref name="checksums"/> available within <paramref name="scope"/> from the
+        /// storage.
         /// </summary>
-        public async ValueTask<IReadOnlyDictionary<Checksum, SolutionAsset>> GetAssetsAsync(int scopeId, IEnumerable<Checksum> checksums, CancellationToken cancellationToken)
+        public static async ValueTask<IReadOnlyDictionary<Checksum, SolutionAsset>> GetAssetsAsync(
+            Scope scope, IEnumerable<Checksum> checksums, CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/SolutionAssetStorage.cs:142]: note: tehse methods becoming static is very suspect in terms of where this code should live.  I could consider making it instance methods on Scope itself.

File: src/Workspaces/Core/Portable/Serialization/SolutionReplicationContext.cs [hunk 3]
```diff
@@ -4,18 +4,18 @@
...
 namespace Microsoft.CodeAnalysis.Serialization
 {
     internal readonly struct SolutionReplicationContext : IDisposable
     {
-        private readonly ArrayBuilder<IDisposable> _resources;
+        private static readonly ObjectPool<ConcurrentSet<IDisposable>> s_pool = new(() => new());
 
-        private SolutionReplicationContext(ArrayBuilder<IDisposable> resources)
-            => _resources = resources;
+        private readonly ConcurrentSet<IDisposable> _resources;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Serialization/SolutionReplicationContext.cs:15]: i was basically not at all confident that the old arraybuilder was not going to be used in a concurrent fashion.  So i moved to a concurrentset to be ultra paranoid about this being safe.

File: src/Workspaces/Remote/Core/SolutionAssetStorage.cs [hunk 4]
```diff
@@ -33,15 +33,18 @@ internal partial class SolutionAssetStorage
         /// the same storage here so that all OOP calls can safely call back into us and get the assets they need, even
         /// if individual calls get canceled.
         /// </summary>
-        private readonly Dictionary<Checksum, (int refCount, Scope scope)> _checksumToRefCountedScope = new();
+        private readonly Dictionary<Checksum, Scope> _checksumToScope = new();
 
-        /// <summary>
-        /// Map from solution checksum to its associated <see cref="SolutionState"/>.
-        /// </summary>
-        private readonly ConcurrentDictionary<Checksum, (SolutionState Solution, SolutionReplicationContext ReplicationContext)> _solutionStates = new(concurrencyLevel: 2, capacity: 10);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/SolutionAssetStorage.cs:41]: Merged the two dictionaries together.  Moved 'refCount' into 'Scope' itself.

File: src/Workspaces/Remote/Core/SolutionAssetStorage.cs [hunk 5]
```diff
@@ -13,121 +13,115 @@
...
 
-            replicationContext.Dispose();
+            // Last ref went away, update our maps while under the lock, then cleanup its context data outside of the lock.
+            _checksumToScope.Remove(checksum);
         }
 
+        scope.ReplicationContext.Dispose();
+    }
+
+    internal partial class Scope
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/SolutionAssetStorage.cs:110]: these helpers moved to be isntance methods on Scope.  They weren't using anything else.   SolutionAssetStorage is basically just a holder for the inflight operations (scopes) for a particular solution checksum.

Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/Core/SolutionAssetStorage.cs:110]: i did this inline in this file with a partial tyepe to make the diff easy to read.  if you are ok with this @dibarbet  i will move this code into the other part.

Comment (dibarbet) [reply src/Workspaces/Remote/Core/SolutionAssetStorage.cs:110]: yup makes sense to me!

**Review by dibarbet — APPROVED** [review]

**Pull request merged**